README
Class-Variable - Perl implementation of class variables with access restrictions
================================================================================

    Documentation on CPAN:
        http://search.cpan.org/~hurricup/Class-Variable/lib/Class/Variable.pm
        
    Documentation on MetaCPAN
        https://metacpan.org/pod/Class::Variable

    CPAN Testers reports
        http://www.cpantesters.org/distro/C/Class-Variable.html

    Testers matrix
        http://matrix.cpantesters.org/?dist=Class-Variable
        
    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Class-Variable

    CPAN Ratings
        http://cpanratings.perl.org/d/Class-Variable

INSTALLATION
============

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE
=====================

Copyright (C) 2015 by Alexandr Evstigneev (hurricup@evstigneev.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


Changes
Revision history for Perl extension Class::Variable.

1.002
	- fixed problems with 5.21 warnings 'redundant'

1.000
	- first release

0.01  Fri Jan 23 20:46:00 2015
	- original version; created by h2xs 1.23 with options
		-AX Class::Variable



t/Bar.pm
package Bar;
use strict;use warnings;
use parent 'Foo';
use Class::Variable;

public 'public2';
protected 'protected2';
private 'private2';

sub get_protected_bar
{
    return shift->protected2;
}
sub set_protected_bar
{
    my( $self, $value) = @_;
    $self->protected2 = $value;
}

sub get_private_bar
{
    return shift->private2;
}
sub set_private_bar
{
    my( $self, $value) = @_;
    $self->private2 = $value;
}

sub get_protected_foo_bar
{
    return shift->protected1;
}
sub set_protected_foo_bar
{
    my( $self, $value) = @_;
    $self->protected1 = $value;
}

sub get_private_foo_bar
{
    return shift->private1;
}
sub set_private_foo_bar
{
    my( $self, $value) = @_;
    $self->private1 = $value;
}


1;

t/Foo.pm
package Foo;
use strict;use warnings;
use Class::Variable;

sub new{return bless {}, shift;}

public 'public1';
protected 'protected1';
private 'private1';

sub get_protected_foo
{
    return shift->protected1;
}
sub set_protected_foo
{
    my( $self, $value) = @_;
    $self->protected1 = $value;
}

sub get_private_foo
{
    return my $var = shift->private1;
}
sub set_private_foo
{
    my( $self, $value) = @_;
    $self->private1 = $value;
}

sub benchmark
{
    my $self = shift;
    require Benchmark;
    Benchmark->import(':all');
    
    my $var = 100;
    timethese( 1000000, {
        '1. Direct write    ' => sub{ $self->{'var'} = $var; },
        '2. Direct read     ' => sub{ $var = $self->{'var'}; },
        '3. Public write    ' => sub{ $self->public1 = $var; },
        '4. Public read     ' => sub{ $var = $self->public1; },
        '5. Protected write ' => sub{ $self->protected1 = $var; },
        '6. Protected read  ' => sub{ $var = $self->protected1; },
        '7. Private write   ' => sub{ $self->private1 = $var; },
        '8. Private read    ' => sub{ $var = $self->private1; },
    });
}

1;


Makefile.PL
use 5.008;use strict;use warnings FATAL => 'all';use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Class::Variable',
    AUTHOR              => 'Alexandr Evstigneev <hurricup@evstigneev.com>',
    VERSION_FROM        => 'lib/Class/Variable.pm', 
    ABSTRACT_FROM       => 'lib/Class/Variable.pm', 
    MIN_PERL_VERSION    => 5.008,
    LICENSE             => 'MIT',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::Harness' => 0,
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Carp' => 0,
        'Exporter' => 0,
        'parent' => 0,
        'Scalar::Util' => 0,
    }, 
    META_MERGE        => {
        'meta-spec' => { 
            'version' => 2 
        },
        'resources' => {
            'license' => [
                'http://www.opensource.org/licenses/mit-license.php',
            ],
            'bugtracker' => {
                'web' => 'https://github.com/hurricup/Class-Variable/issues'
            },
            'repository' => {
                'type' => 'git',
                'url' => 'git://github.com/hurricup/Class-Variable.git',
                'web' => 'https://github.com/hurricup/Class-Variable',
            },
        },
    },    
);


lib/Class/Visibility.pm
package Class::Visibility;
use strict; use warnings FATAL => 'all'; 
use parent 'Exporter';
use 5.000000;
use Carp;
use Scalar::Util 'weaken';

our $VERSION = '1.000000';

our @EXPORT;

my $NS = {};

push @EXPORT, 'public';
sub public
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_public_variable($package, $name);
    }
}

push @EXPORT, 'protected';
sub protected
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_protected_variable($package, $name);
    }
}

push @EXPORT, 'private';
sub private
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_private_variable($package, $name);
    }
}

sub get_public_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        $NS->{$self}->{$name};
    };
}

sub get_protected_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        croak sprintf(
            "Access violation: protected variable %s of %s available only to class or subclasses, but not %s."
            , $name
            , $package
            , caller ) if not caller->isa($package);
            
        $NS->{$self}->{$name};
    };
}

sub get_private_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        croak sprintf(
            "Access violation: private variable %s of %s available only to class itself, not %s."
            , $name
            , $package
            , caller ) if caller ne $package;
            
        $NS->{$self}->{$name};
    };
}

1;
__END__
=head1 NAME

Class::Visibility - Perl implementation of class variables visibility limitation.

=head1 SYNOPSIS

  use Class::Visibility;

=head1 DESCRIPTION


=head1 BUGS AND IMPROVEMENTS

If you found any bug and/or want to make some improvement, feel free to participate in the project on GitHub: L<https://github.com/hurricup/Class-Visibility>

=head1 LICENSE

This module is published under the terms of the MIT license, which basically means "Do with it whatever you want". For more information, see the LICENSE file that should be enclosed with this distributions. A copy of the license is (at the time of writing) also available at L<http://www.opensource.org/licenses/mit-license.php>.

=head1 SEE ALSO

=over

=item * Main project repository and bugtracker: L<https://github.com/hurricup/Class-Visibility>

=item * See also: L<Class::Property>. 

=back

=head1 AUTHOR

Copyright (C) 2015 by Alexandr Evstigneev (L<hurricup@evstigneev.com|mailto:hurricup@evstigneev.com>)



=cut


lib/Class/Variable.pm
package Class::Variable;
use 5.008;
use strict; use warnings FATAL => 'all'; 
use Exporter 'import';
use Carp;
use Scalar::Util 'weaken';

our $VERSION = '1.002'; # <== update version in pod

our @EXPORT;

my $NS = {};

push @EXPORT, 'public';
sub public($;)
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_public_variable($package, $name);
    }
}

push @EXPORT, 'protected';
sub protected($;)
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_protected_variable($package, $name);
    }
}

push @EXPORT, 'private';
sub private($;)
{
    my @names = @_;
    my $package = (caller)[0];
    foreach my $name (@names)
    {
        no strict 'refs';
        *{$package.'::'.$name } = get_private_variable($package, $name);
    }
}

sub get_public_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        $NS->{$self}->{$name};
    };
}

sub get_protected_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        croak sprintf(
            "Access violation: protected variable %s of %s available only to class or subclasses, but not %s."
            , $name || 'undefined'
            , $package || 'undefined'
            , caller()  || 'undefined' ) if not caller()->isa($package);
            
        $NS->{$self}->{$name};
    };
}

sub get_private_variable($$)
{
    my( $package, $name ) = @_;
    
    return sub: lvalue
    {
        my $self = shift;
        if( 
            not exists $NS->{$self}
            or not defined $NS->{$self}->{' self'} 
        )
        {
            $NS->{$self} = {
                ' self' => $self
            };
            weaken $NS->{$self}->{' self'};
        }
        
        croak sprintf(
            "Access violation: private variable %s of %s available only to class itself, not %s."
            , $name || 'undefined'
            , $package || 'undefined'
            , caller()  || 'undefined' ) if caller() ne $package;
            
        $NS->{$self}->{$name};
    };
}


1;
__END__
=head1 NAME

Class::Variable - Perl implementation of class variables with access restrictions.

=head1 VERSION

Version 1.002

=head1 SYNOPSIS

This module allows You to create class members with access restrictions, using intuitive syntax:

    package Foo;
    use Class::Variable;
    
    public      'var1', 'var2';   # these variables available everywhere 
    protected   'var3', 'var4';   # these variables available only in objects of Foo class or subclasses
    private     'var5', 'var6';   # these variables available only in objects of Foo class 
    
meanwhile somewhere else ...
    
    use Foo;
    
    my $foo = Foo->new();
    
    $foo->var1 = "Public var content";      # works fine
    $foo->var3 = "Protected var content";   # croaks, protected
    $foo->var5 = "Private var content";     # croaks, private
    
All generated class variables are actually lvalue methods and can be inherited by subclasses.
    
=head1 DESCRIPTION

Module exports three methods, required to define variables: C<public>, C<protected> and C<private>.

Internally, there is a namespace variable in C<Class::Variable> package, which is not available from the outside and contains all data per object, using weak references to avoid duplicated references (not sure if it's possible). 

Generated class variables are lvalue subs with access control in them.

Don't forget, that data from generated variables is not encapsulated in object and can't be serialized.

=head1 BENCHMARKS

Here is a comparision of direct acces to hash elements and access to generated variables:

    1. Direct write    :  1 wallclock secs ( 0.58 usr +  0.00 sys =  0.58 CPU) @ 17331022.53/s (n=10000000)
    2. Direct read     :  1 wallclock secs ( 0.56 usr +  0.00 sys =  0.56 CPU) @ 17793594.31/s (n=10000000)
    3. Public write    : 11 wallclock secs (10.36 usr +  0.00 sys = 10.36 CPU) @ 965437.34/s (n=10000000)
    4. Public read     : 10 wallclock secs (10.37 usr +  0.00 sys = 10.37 CPU) @ 963948.33/s (n=10000000)
    5. Protected write : 15 wallclock secs (14.21 usr +  0.00 sys = 14.21 CPU) @ 703630.73/s (n=10000000)
    6. Protected read  : 14 wallclock secs (14.13 usr +  0.00 sys = 14.13 CPU) @ 707513.80/s (n=10000000)
    7. Private write   : 12 wallclock secs (11.31 usr +  0.00 sys = 11.31 CPU) @ 884173.30/s (n=10000000)
    8. Private read    : 11 wallclock secs (11.23 usr +  0.00 sys = 11.23 CPU) @ 890313.39/s (n=10000000)

We can see, that public variables works 18 times slower, than direct access, protected variables are 25 times slower, than direct acces and private variables are 20 times slower, than direct access.
    
=head1 SEE ALSO

=over

=item * Main project repository and bugtracker: L<https://github.com/hurricup/Class-Variable>

=item * CPAN Testers reports: L<http://www.cpantesters.org/distro/C/Class-Variable.html>

=item * Testers matrix: L<http://matrix.cpantesters.org/?dist=Class-Variable>
        
=item * AnnoCPAN, Annotated CPAN documentation: L<http://annocpan.org/dist/Class-Variable>

=item * CPAN Ratings: L<http://cpanratings.perl.org/d/Class-Variable>

=item * See also: L<Class::Property>, L<Class::Accessor::Lazy>. 

=back

=head1 LICENSE

This module is published under the terms of the MIT license, which basically means "Do with it whatever you want". For more information, see the LICENSE file that should be enclosed with this distributions. A copy of the license is (at the time of writing) also available at L<http://www.opensource.org/licenses/mit-license.php>.

=head1 AUTHOR

Copyright (C) 2015 by Alexandr Evstigneev (L<hurricup@evstigneev.com|mailto:hurricup@evstigneev.com>)

=cut
